src/main/java/com/thealgorithms/sorts/SmoothSort.java

```diff
@@ -0,0 +1,227 @@
+package com.thealgorithms.sorts;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+
+/**
+ * Wikipedia: https://en.wikipedia.org/wiki/Smoothsort
+ */
+public final class SmoothSort implements SortAlgorithm {
+
+    public SmoothSort() {
+    }
+
+    private static Integer[] getLeonardoNumbers() {
+        Integer[] leonardoNumbers = {1, 1, 3, 5, 9, 15, 25, 41, 67, 109, 177, 287, 465, 753, 1219, 1973, 3193, 5167, 8361, 13529, 21891, 35421, 57313, 92735, 150049, 242785, 392835, 635621, 1028457, 1664079, 2692537, 4356617, 7049155, 1405773, 18454929, 29860703, 48315633, 78176337, 126491971,
+            204668309, 331160281, 535828591};
+
+        return leonardoNumbers;
+    }
+
+    private static <T extends Comparable<T>> void smoothSort(T[] array) {
+        int length = array.length;
+        int leonardoHeapSize = 0; // start with size 0
+        int leonardoLevelTracker = 0; // No leonardo tree present initially
+
+        while (leonardoHeapSize < length) {
+            // if two trees with consequtive level
+            // combine them to get new tree
+            // else if there is no Level 1, add the node as level 1
+            // else add the node as Level 0
+            // perform shiftRoot and restore heap property
+
+            Integer[] consecutiveTreeIndices = findConsecutiveLeonardoTreeIndices(leonardoLevelTracker);
+            if (consecutiveTreeIndices[0] != -1) {
+                // if 0th or 1st index is -1 that implies there are no concequtive trees
+                leonardoLevelTracker = leonardoLevelTracker & ~(1 << consecutiveTreeIndices[0]);
+                leonardoLevelTracker = leonardoLevelTracker & ~(1 << consecutiveTreeIndices[1]);
+                leonardoLevelTracker = leonardoLevelTracker | (1 << consecutiveTreeIndices[1] + 1);
+            } else if ((leonardoLevelTracker & 2) == 0) {
+                leonardoLevelTracker = leonardoLevelTracker | (1 << 1);
+            } else {
+                leonardoLevelTracker = leonardoLevelTracker | (1 << 0);
+            }
```

Comment:
Is it possible to apply the [`SingleBitOperations`](https://github.com/TheAlgorithms/Java/blob/a710fe11c47bdfb80342f851c189d167035b7b1c/src/main/java/com/thealgorithms/bitmanipulation/SingleBitOperations.java) (or something from [`bitmanipulation`](https://github.com/TheAlgorithms/Java/tree/master/src/main/java/com/thealgorithms/c))?

Reply:
I did not come accross this class earlier. 
Yes, this class can definately be used. I shall modify this section of the code as well.